internal/pkg/manifest/svc.go

```diff
@@ -39,7 +40,85 @@ type Service struct {
 
 // ServiceImage represents the service's container image.
 type ServiceImage struct {
-	Build *string `yaml:"build"` // Path to the Dockerfile.
+	Build BuildArgsOrString `yaml:"build"` // Path to the Dockerfile.
+}
+
+// Dockerfile returns the path to the service's Dockerfile, defaulting to
+// the filename Dockerfile if not otherwise specified.
+func (s *ServiceImage) Dockerfile() string {
+	// Prefer to use the "dockerfile" string in BuildArgs. Otherwise,
+	// use "Context/Dockerfile", then "BuildString", then "Dockerfile"
+	if s.Build.BuildArgs.Dockerfile != nil {
+		return aws.StringValue(s.Build.BuildArgs.Dockerfile)
+	}
+
+	var dfPath string
+	if s.Build.BuildString != nil {
+		dfPath = aws.StringValue(s.Build.BuildString)
+	} else if s.Build.BuildArgs.Context != nil {
+		// This is a degraded experience, should only seldom happen when Dockerfile isn't explicit.
+		dfPath = aws.StringValue(s.Build.BuildArgs.Context)
+		dfPath = filepath.Join(dfPath, "Dockerfile")
+	} else {
+		dfPath = "./Dockerfile"
+	}
+
+	return dfPath
+}
+
+// Context returns the build context directory if it exists, otherwise an empty string.
+func (s *ServiceImage) Context() string {
+	return aws.StringValue(s.Build.BuildArgs.Context)
+}
+
+// Args returns the args section, if it exists, to override args in the dockerfile.
+// Otherwise it returns an empty map.
+func (s *ServiceImage) Args() map[string]string {
+	return s.Build.BuildArgs.Args
+}
+
+// BuildArgsOrString is a custom type which supports unmarshaling yaml which
+// can either be of type string or type DockerBuildArgs.
+type BuildArgsOrString struct {
+	BuildString *string
+	BuildArgs   DockerBuildArgs
+}
+
```

Comment:
What do you think about implementing the [`yaml.Marshaler`](https://godoc.org/gopkg.in/yaml.v3#Marshaler) interface? So that if `BuildString` is not nil then returns the string, otherwise it returns `yaml.Marshal(b.BuildArgs)`

Reply:
Going to leave this as is for the time being as discussed, since we use `template.Execute` instead of `yaml.Marshal` to create our manifests.